### ISSUE TITLE
FEATURE REQUEST: Implementing Tunnel Creation without sudo

### ISSUE THREAD

**Issue by myungkyojung** [doronz88/pymobiledevice3#1260]
Can you implement a feature that allows creating a tunnel without using sudo? 
In a similar way, in go-ios, the --userspace option can be added when creating a tunnel to start it without sudo.

**Comment by doronz88** [comment]
Implementing this in python would really be practical if it's done using native which would be hard to write and maintain. You're welcome to write such a module yourself and submit a PR of using it

**Comment by JJTech0130** [comment]
This would be closed if the current solution we are discussing in https://github.com/doronz88/pymobiledevice3/issues/566 (borrowing go-ios's userspace implementation) was implemented

**Comment by ghost** [comment]
Created a [fork](https://github.com/xunmod/pymobiledevice3) to make pymobiledevice3 borrow go-ios's userspace TUN implementation  
`go-ios tunnel start` and `go-ios tunnel ls` should show something like this:
```
[
  {
    "address": "fdb6:a01e:9c0::1",
    "rsdPort": 52516,
    "udid": "[UDID]",
    "userspaceTun": true,
    "userspaceTunPort": 60108
  }
]
```
Then `pymobiledevice3 developer dvt ls / --rsd fdb6:a01e:9c0::1 52516 --userspace-wrapper 127.0.0.1 60108`

==========
Or just use `pymobiledevice3 developer dvt ls / --go-ios ''` to auto discover devices from go-ios (requires `go-ios tunnel start` running)

**Comment by doronz88** [comment]
@xunmod though I see the benefit of using this approach, both deployment and the complexity of all the available options to connect to the device are becoming too much on my opinion. If `go-ios` has something like tunneld that automatically searches for device I think the best approach would be to just adapt to his API instead so pymobiledevice3 will work with both implementations seamlessly.

**Comment by ghost** [comment]
> [@xunmod](https://github.com/xunmod) though I see the benefit of using this approach, both deployment and the complexity of all the available options to connect to the device are becoming too much on my opinion. If `go-ios` has something like tunneld that automatically searches for device I think the best approach would be to just adapt to his API instead so pymobiledevice3 will work with both implementations seamlessly.

`go-ios` do have one and I call it `agent`, which can be used by the `--go-ios` parameter. Should I delete the `--userspace-wrapper` parameter?

**Comment by doronz88** [comment]
Yeah. I think its preferable to only rely on a single tunneld protocol

**Comment by myungkyojung** [comment]
any update???

**Comment by jkcoxson** [comment]
For the record, I've implemented a software TCP stack into my [library](https://github.com/jkcoxson/idevice), I don't think it would be difficult to port to this project. Only roughly 300 lines of TCP handling code.

**Comment by doronz88** [comment]
PRs are welcome

**Comment by doronz88** [comment]
Closing — creating a tunnel without sudo is now supported via the no-root `--userspace` tunnel (iOS 17+).
